src/disco/gui/fd_gui_printf.c

```diff
@@ -0,0 +1,1172 @@
+
+#include "fd_gui_printf.h"
+
+#include "../../ballet/http/fd_http_server_private.h"
+#include "../../ballet/utf8/fd_utf8.h"
+
+#define FD_GUI_PRINTF_DEBUG 0
+
+static void
+jsonp_strip_trailing_comma( fd_gui_t * gui ) {
+  if( FD_LIKELY( !gui->http->stage_err &&
+                 gui->http->stage_len>=1UL &&
+                 gui->http->oring[ (gui->http->stage_off%gui->http->oring_sz)+gui->http->stage_len-1UL ]==(uchar)',' ) ) {
+    gui->http->stage_len--;
+  }
+}
+
+static void
+jsonp_open_object( fd_gui_t *   gui,
+                   char const * key ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":{", key );
+  else                   fd_http_server_printf( gui->http, "{" );
+}
+
+static void
+jsonp_close_object( fd_gui_t * gui ) {
+  jsonp_strip_trailing_comma( gui );
+  fd_http_server_printf( gui->http, "}," );
+}
+
+static void
+jsonp_open_array( fd_gui_t *   gui,
+                  char const * key ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":[", key );
+  else                   fd_http_server_printf( gui->http, "[" );
+}
+
+static void
+jsonp_close_array( fd_gui_t * gui ) {
+  jsonp_strip_trailing_comma( gui );
+  fd_http_server_printf( gui->http, "]," );
+}
+
+static void
+jsonp_ulong( fd_gui_t *   gui,
+             char const * key,
+             ulong        value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%lu,", key, value );
+  else                   fd_http_server_printf( gui->http, "%lu,", value );
+}
+
+static void
+jsonp_long( fd_gui_t *   gui,
+            char const * key,
+            long         value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%lu,", key, value );
+  else                   fd_http_server_printf( gui->http, "%ld,", value );
+}
+
+static void 
+jsonp_double( fd_gui_t *   gui,
+              char const * key,
+              double       value ) {
+  if( FD_LIKELY( key ) ) fd_http_server_printf( gui->http, "\"%s\":%.2f,", key, value );
+  else                   fd_http_server_printf( gui->http, "%.2f,", value );
+}
+
+static void
+jsonp_sanitize_str( fd_http_server_t * http,
+                    ulong              start_len ) {
+  /* escape quotemark, reverse solidus, and control chars U+0000 through U+001F
+     just replace with a space */
+  uchar * data = http->oring;
+  for( ulong i=start_len; i<http->stage_len; i++ ) {
+    if( FD_UNLIKELY( data[ (http->stage_off%http->oring_sz)+i ] < 0x20 ||
+                     data[ (http->stage_off%http->oring_sz)+i ] == '"' ||
+                     data[ (http->stage_off%http->oring_sz)+i ] == '\\' ) ) {
+      data[ (http->stage_off%http->oring_sz)+i ] = ' ';
+    }
+  }
+}
+
+FD_FN_UNUSED static void
+print_char_buf_to_buf_escape_hex(const char *buf, ulong buf_sz, char *out_buf, ulong out_buf_sz) {
+  /* Prints invisible chars as escaped hex codes */
+  ulong idx = 0;
+  for( ulong i = 0; i < buf_sz; ++i ) {
+    uchar c = (uchar)buf[i];
```

Comment:
Not sure why this is cast to `uchar`. `isprint` takes an int. You can just keep it as a `char`